compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -609,7 +611,12 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
             .max()
             .unwrap_or(ty::UniverseIndex::ROOT);
 
-        Canonical { max_universe, variables: canonical_variables, value: (base.value, out_value) }
+        Canonical {
```

Comment:
```suggestion
        assert!(!infcx.is_some_and(|infcx| infcx.defining_opaque_types != base.defining_opaque_types))
        Canonical {
```
This assertion would've caught this bug.